### ISSUE TITLE
Conditional Validation

### ISSUE THREAD

**Issue by geekravi92** [express-validator/express-validator#439]
Hello, technocrats!

I'm facing a hard time implementing conditional validation. As per the new API version, validation checks are passed in the form of an array as a second parameter, well, this is what I could infer, to the routing methods. Lemme give you a scenario to make it clear. 

I have a login page with two tabs defining Student Login and Teacher Login. Each tab has a Bootstrap Form with Student ID (or Teacher ID) input field, a Password input field, a hidden input with a fixed value for each (say SID for student and TID for teacher),  and a submit button. The POST method for the page is defined in index.js. 
Now, the issue is I need to check the value of the hidden input first and on the basis of that only I need to choose the right route for the student and the teacher. I need to check all the input validations also for respective fields in Student Login form and Teacher Login form. It's just like dealing with two forms in one POST method. Where should I be writing the conditional code for the route?

Any help is heartily appreciated.

Thank you
Arvind Jaiswal

**Comment by gustavohenke** [comment]
If I understood you correctly, you can use `oneOf`:

```js
app.post('/form', oneOf([
  check('hidden').equals('SID'),
  check('studentId').exists()
], [
  check('hidden').equals('TID'),
  check('teacherId').exists()
], 'a message'), check('password').exists(), routeHandler);
```

**Comment by geekravi92** [comment]
This is not working.

**Comment by gustavohenke** [comment]
Part of what you need will be achievable when #434 gets implemented.
Basically, the idea there is to add support for validation styles like this:

```js
oneOf([[
  check('idType').equals('SID'),
  check('id').someValidator()
], [
  check('idType').equals('TID'),
  check('id').someValidator()
]])
```

But, unfortunately, the support for grouping validations together within `oneOf` is not here yet.
For now, what I can suggest you is to split the routes, eg `/login/student` and `/login/teacher`.

**Comment by thomasdashney** [comment]
While `oneOf` does get the job done here, it will emit excessive errors. Using the above example, if we only care about first set of validations (ie. when `idType` is `SID`), and that fails, you will get error messages for `idType` not equalling `TID`.

**Comment by thomasdashney** [comment]
Maybe could look something like this:
```js
const { checkIf, check } = require('express-validator')

[
  checkIf(({ req }) => req.body.idType === 'SID', [
    check('id').someValidator()
  ]),
  checkIf(({ req }) => req.body.idType === 'TID', [
    check('id').someValidator()
  ])
]
```

**Comment by JeffMatthews2** [comment]
I am encountering the same issue.  I check to see if userName is empty and if userName is not already registered in the database.  I don't want to check the database if userName is empty.  If userName is **not** empty, then, and only then, do I want to check the database.

Is there are way to do this?

**Comment by JeffMatthews2** [comment]
I was able to simulate conditional validation by rechecking my prior validation rules in my custom validator, and if they fail to force the custom to pass without inspecting the database.

```
          check('name')

		.trim()

		.isLength( { min: 5, max: undefined } )
		.withMessage('Name must be at least 5 characters.')

		.custom(name => {
			if (name.trim().length >=5) {
				return new Promise( (resolve, reject) => {
					User.getUserByName(name).then(user => {
						reject('This name has already been registered');
					}).catch( err => {
						resolve(true);
					});
				});
			} else {
				return true;
			}

		}),


		sanitize('name').trim()

], (req, res, next) => {

```

**Comment by neobyteUK** [comment]
@thomasdashney @JeffMatthews2 you might be interested in #495?

**Comment by benmerchant** [comment]
How could I check if a field exists, and if it doesn't, I'm not worried about it, but if it is, I'd like to validate it. For instance, I have a dynamic table who's inputs are enabled/disabled by checkboxes. If a box is unchecked, the corresponding input is attr('disabled', true), so it won't POST any data. However, if it is enabled, I would like to be able to validate it. 

![express-validator-question_number_439](https://user-images.githubusercontent.com/22625009/37188488-6748704c-231d-11e8-814d-13ca79088777.PNG)

**Comment by gustavohenke** [comment]
Hi @benmerchant, you might be looking at [`optional`](https://github.com/ctavan/express-validator#optionaloptions).

**Comment by benmerchant** [comment]
Thank you, @gustavohenke. That's exactly it!
I seriously just did this for testing overall flow through the app.
var query = Employee.findOneAndUpdate({_id:req.params.id},req.body);

**Comment by bulicmatko** [comment]
Hi @ctavan, @gustavohenke 

Here's a situation;

Let's say we have a form with two dependent fields;
- radio list with two options: Call required - 'Yes' : 'No'
- and phone number field

Phone number should be required only in case user selected that he wants to be called;
If other option is selected, `phoneNumber` field is not required but is still validated (i.e. `isNumeric()`).

I haven't been able to find a way to do this kind of validation and keep `errors.mapped()` "nice and tidy".

If using `oneOf`, `errors.mapped()` returns `_errors` key which have to be manually mapped.
The idea is to keep error messages the same way as they are without using `oneOf`.

Maybe passing a function that returns validation options can be easily implemented.
This could solve the problem with field dependent validation. 

```
body('phoneNumber')
  .isLength(body => body.phoneCallRequired === 'yes' ? { min: 1 } : { min: 0 })
  .withMessage('Phone number is required')
  .trim()
```

Thanks

**Comment by pabloromeu** [comment]
Hi all, 

I have found a solution for this issue. The point is to select which middleware is going to be executed on your route but when the request starts. Then you can use this code:

```
const async = require("async");

function dynamicMiddleware(validators, req,res,next) {
    async.eachSeries(validators, function(middleware,doneMiddleware) {
        middleware.bind(null,req,res,doneMiddleware)()
    }, function(err) {
        if (err) {
            return res.status(500).json({error: "there was a problem with your middleware"});
        }
        else {
            next(err);
        }
    })
}

router.route('/track')
     .post( // Some middleware here
             function (request, response, next) {
            let validatorsToSend =[];
            switch (request.body.accountType)){ // Validators executed dynamically on the request
                case "individual":
                    validatorsToSend = [
                        body('firstName', 'No valid firstName provided').optional({nullable:false}).isLength({min: 1}),
                        body('surname', 'No valid surname provided').optional({nullable:false}).isLength({min: 1}),
                    ];
                    break;
                case "corporate":
                    validatorsToSend = [
                        body('companyName', 'No valid companyName provided').optional({nullable:false}).isLength({min: 1}),
                    ];
                    break;
                default:
                    return response.status(400).json();
            }

            dynamicMiddleware(validatorsToSend,request,response,next);
        },

```

I hope this helps!

**Comment by gustavohenke** [comment]
Correct @formatically it isn't yet.

Folks can refer to #658 to see the current proposal; the API is exactly the one suggested by @thomasdashney.

**Comment by andhikamaheva** [comment]
That's why I'm still using Legacy API just because I can set conditional validations easily 🙄

**Comment by gustavohenke** [comment]
👋 this has been published to [`v6.1.0`](https://github.com/express-validator/express-validator/releases/tag/v6.1.0) 🚢
